src/VisualStudio/Next/Remote/RemoteHostClientServiceFactory.RemoteHostClientService.cs

```diff
@@ -0,0 +1,178 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.Diagnostics;
+using Microsoft.CodeAnalysis.Execution;
+using Microsoft.CodeAnalysis.Extensions;
+using Microsoft.VisualStudio.LanguageServices.Implementation.Remote;
+
+namespace Microsoft.VisualStudio.LanguageServices.Remote
+{
+    internal partial class RemoteHostClientServiceFactory
+    {
+        private class RemoteHostClientService : IRemoteHostClientService
+        {
+            private readonly Workspace _workspace;
+            private readonly IDiagnosticAnalyzerService _analyzerService;
+
+            private readonly object _gate;
+
+            private CancellationTokenSource _shutdown;
+            private Task<RemoteHostClient> _instance;
+
+            public RemoteHostClientService(Workspace workspace, IDiagnosticAnalyzerService analyzerService)
+            {
+                _gate = new object();
+
+                _workspace = workspace;
+                _analyzerService = analyzerService;
+            }
+
+            public void Enable()
+            {
+                lock (_gate)
+                {
+                    if (_instance != null)
+                    {
+                        // already enabled
+                        return;
+                    }
+
+                    if (!_workspace.Options.GetOption(RemoteHostOptions.RemoteHost))
+                    {
+                        // not turned on
+                        return;
+                    }
+
+                    // make sure we run it on background thread
+                    _shutdown = new CancellationTokenSource();
+
+                    var token = _shutdown.Token;
+                    _instance = Task.Run(() => EnableAsync(token), token);
+                }
+            }
+
+            public void Disable()
+            {
+                lock (_gate)
+                {
+                    if (_instance == null)
+                    {
+                        // already disabled
+                        return;
+                    }
+
+                    var instance = _instance;
+                    _instance = null;
+
+                    RemoveGlobalAssets();
+
+                    _shutdown.Cancel();
+
+                    try
+                    {
+                        instance.Result.Shutdown();
```

Comment:
Does instance.Result throw a cancellation exception in the canceled state?  Or will you get an AggregateException?  I think you get the latter.

Instead, you can do instance.Wait(_shutdown.Token) instead.